lib/fastlane/plugin/wpmreleasetoolkit/actions/common/buildkite_add_trigger_step_action.rb

```
@@ -0,0 +1,165 @@
+# frozen_string_literal: true
+
+require 'open3'
+
+module Fastlane
+  module Actions
+    class BuildkiteAddTriggerStepAction < Action
+      def self.run(params)
+        # Extract parameters
+        pipeline_file = params[:pipeline_file]
+        build_name = File.basename(pipeline_file, '.yml')
+        message = params[:message] || build_name
+        branch = params[:branch] || `git rev-parse --abbrev-ref HEAD`.strip
```

I love the convenience and conciseness of using the backticks, but part of me says we'd be better off using the [Git gem](https://github.com/ruby-git/ruby-git), specially thinking about error handling 🤔

---

To be honest I've been pulling back a bit about using the Git gem in release-toolkit, because since we started using it I've noticed a couple of things in its implementation details that caused issues.

In particular:
 - Using the Git gem to check if a branch exists locally has a side effect of the gem fetching and catching the list of all the branches in the repo, local and remote, before checking if the branch we asked about was in the array of cached branches
 - Which in some repos can take some time, at least more time than just checking if a local branch of a specific given name exists with a simple `git branch --list "#{branch_name}"`
 - And also causes errors in some cases because the implementation of the Git gem to parse the list of branches from `git branch --list --all` to cache the whole list uses some RegEx string parsing, and some of those fail to account for special character in branch names… including spaces.
 - We actually encountered errors not that long ago about some release automation failing just because it happened that one of the remote branches in a repo contained a space, even if that branch had nothing to do at all with what the release process was manipulating.

So after those experiences, performance issues, and bugs, my trust in this Git gem has lowered a bit since we first started considering it 😓

---

> and some of those fail to account for special character in branch names… including spaces

Indeed, I've seen it happening as well! I totally understand the losing of confidence in the lib.

I think it's really nice that the Git gem abstracts the details of git calls and the arguments... but at the same time we lose control and it might yield commands or have an internal implementation that might have side effects we don't know / cannot control.

One of my main worries is error handling -- I think using backticks will just "swallow" any eventual error? We can always also use `sh` (which is nice when we have arguments) but for simple cases it may be not strictly needed?

I'm fine on deciding to go in either direction as long as we're sure about the trade offs and keep it consistent 👍

---

Yeah that's a fair point. And actually my initial implementation tried to use `Open3.capture3` to run that command just like I use for running the `buildkite-agent` command… but then that overcomplicated the code and I figured it wasn't worth it.

I like the idea of using `sh` though. iinm we can even provide options to Fastlane's `sh` helper to tell it not to print the details of such shell command call in the output nor consider that call as being an action when it prints the summary table at the end of a lane… will check if that works.

---

Switched to using `sh` in `35685537`, and tested how it looks in a real world use case via https://github.tumblr.net/Tumblr/ios/pull/29788/commits/147498ea3cd8f74aa3eddd762a1f018f7678d919 and https://buildkite.com/automattic/tumblr-ios/builds/33625/steps?sid=0196ed6b-2fad-4257-80d5-96ed9e40dab2#0196ed6b-2fe4-4e42-baad-c4a9f64dd40c/212-216:

```
[07:19:38 -0400]: ----------------------------------------
[07:19:38 -0400]: --- Step: buildkite_add_trigger_step ---
[07:19:38 -0400]: ----------------------------------------
[07:19:38 -0400]: $ git rev-parse --abbrev-ref HEAD
[07:19:38 -0400]: ▸ test/AINFRA-98
[07:19:39 -0400]: Added a trigger step to the current Buildkite build to start a new build for fake-beta-build.yml on branch test/AINFRA-98


+-----------------------------------------------------+
|                  fastlane summary                   |
+------+--------------------------------+-------------+
| Step | Action                         | Time (in s) |
+------+--------------------------------+-------------+
| 1    | setup_ci                       | 0           |
| 2    | git checkout -b test/AINFRA-98 | 0           |
| 3    | git_commit                     | 0           |
| 4    | push_to_git_remote             | 1           |
| 5    | is_ci                          | 0           |
| 6    | buildkite_add_trigger_step     | 1           |
+------+--------------------------------+-------------+
```
I guess it's ok for the command and its output to be printed in the logs then, since they are hidden under a collapsible group anyway (thanks to #638) and don't appear in the fastlane summary table 👍